iothub/device/src/Transport/AmqpClientConnectionPool.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.IO;
+    using System.Collections.Concurrent;
+    using Microsoft.Azure.Devices.Shared;
+
+    delegate void RemoveClientConnectionFromPool(DeviceClientEndpointIdentity deviceClientEndpointIdentity);
+
+    /// <summary>
+    /// 
+    /// </summary>
+    internal class AmqpClientConnectionPool
+    {
+        ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection> iotHubSasPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> singleSasPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> sasMuxPool;
+        ConcurrentDictionary<DeviceClientEndpointIdentity ,AmqpClientConnection> x509Pool;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        internal AmqpClientConnectionPool()
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}");
+
+            singleSasPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            x509Pool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            sasMuxPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+            iotHubSasPool = new ConcurrentDictionary<DeviceClientEndpointIdentity, AmqpClientConnection>();
+        }
+
+        internal AmqpClientConnection GetClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpClientConnectionPool)}.{nameof(GetClientConnection)}");
+
+            if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySasSingle)
+            {
+                return GetIoTHubSasSingleClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityX509)
+            {
+                return GetIoTHubX509ClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentitySASMux)
+            {
+                return GetIoTHubSasMuxClientConnection(deviceClientEndpointIdentity);
+            }
+            else if (deviceClientEndpointIdentity is DeviceClientEndpointIdentityIoTHubSas)
+            {
+                return GetIoTHubSasClientConnection(deviceClientEndpointIdentity);
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+        }
+        internal void RemoveClientConnection(DeviceClientEndpointIdentity deviceClientEndpointIdentity)
```

Comment:
Should connection be closed if it's removed from the pool?

Reply:
This function is called from the OnAmqpConnectionClose handler so it is already closed.

---
In reply to: [251669848](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251669848) [](ancestors = 251669848)